draft-ietf-httpbis-message-signatures.md

```diff
@@ -382,7 +343,153 @@ This example shows a canonicalized value for the parameters of a given signature
+~~~
+
+Would result in the following `@method` value:
+
+~~~
+"@method": POST
+~~~
+
+### Authority {#content-request-authority}
+
+The `@authority` identifier refers to the authority component of the target of the HTTP request message. In HTTP 1.1, this is usually conveyed using the `Host` header, while in HTTP 2 and HTTP 3 it is conveyed using the `:authority` pseudo-header. The value is the fully-qualified authority component of the request, comprised of the host and, optionally, port of the request target, as a string.
+
+For example, the following request message:
+
+~~~
+POST /path?param=value HTTP/1.1
+Host: www.example.com
+~~~
+
+Would result in the following `@authority` value:
+
+~~~
+"@authority": www.example.com
+~~~
+
+### Scheme {#content-request-scheme}
+
+The `@scheme` identifier refers to the scheme of the target URL of the HTTP request message. The value is the scheme as a string.
+
+For example, the following request message requested over plain HTTP:
+
+~~~
+POST /path?param=value HTTP/1.1
+Host: www.example.com
+~~~
+
+Would result in the following `@scheme` value:
+
+~~~
+"@scheme": http
+~~~
+
+### Request Origin {#content-request-origin}
+
+The `@request-origin` identifier refers to the full origin component of the HTTP request message. The value is combination of the absolute path and query components of the request URL.
+
+For example, the following request message:
+
+~~~
+POST /path?param=value HTTP/1.1
+Host: www.example.com
+~~~
+
+Would result in the following `@request-origin` value:
+
+~~~
+"@request-origin": /path?param=value
+~~~
+
+The following OPTIONS request message:
+
+~~~
+OPTIONS * HTTP/1.1
+Host: www.example.com
+~~~
+
+Would result in the following `@request-origin` value:
+
+~~~
+"@request-origin": *
+~~~
+
+### Path {#content-request-path}
+
+The `@path` identifier refers to the target path of the HTTP request message. The value is the absolute path of the request target, with no query parameters and no trailing `?` character.
+
+For example, the following request message:
+
+~~~
+POST /path?param=value HTTP/1.1
+Host: www.example.com
+~~~
+
+Would result in the following `@path` value:
+
+~~~
+"@path": /path
+~~~
+
+### Query Components {#content-request-query}
+
+The `@query` identifier refers to the query parameters of the HTTP request message. When the identifier is used with no parameters, the value is the entire normalized query string, not including the leading `?` character.
+
+For example, the following request message:
+
+~~~
+POST /path?param=value&foo=bar&baz=batman HTTP/1.1
+Host: www.example.com
+~~~
+
+Would result in the following `@query` value:
+
+~~~
+"@query": param=value&foo=bar&baz=batman
+~~~
+
+The identifier supports an OPTIONAL `key` parameter containing the name of a single query parameter. When this parameter is present, the value is the normalized key-value pair of the named query parameter, not including any leading `?` characters or separating `&` characters.
```

Comment:
Hmm. This format is specific to HTML, although used in a very wide fashion. At a minimum, we need to mention that this is meaningful when that format is use, with a reference to the appropriate part of url.spec.whatwg.org.

Reply:
Good point. It's wide enough that most developers don't make that distinction, but we should for sure reference the WHATWG spec here specifically. Would a different name help that, like `query-parameters`?

Reply:
Hmm, not sure.

Reply:
While this PR goes in the right direction, I think that we should limit to resource identification (eg. target-uri).

Reply:
@ioggstream We need to consider how people use this in the wild, not just the purity of HTTP semantics. The fact is that people do use HTML query parameters all the time, and the overwhelming majority of developers even consider that to be part of HTTP, not HTML. I do think we need to allow for signing of the query string as-is, but I support signing of query parameters when they exist.